amethyst_ui/src/widgets.rs

```diff
@@ -0,0 +1,98 @@
+use amethyst_core::specs::prelude::{Component, ReadStorage, WriteStorage};
+use derivative::Derivative;
+use std::{
+    ops::Deref,
+    collections::HashMap,
+};
+
+pub trait Widget {
+    fn get_id(&self) -> &str;
+}
+
+#[derive(Derivative)]
+#[derivative(Default(bound = ""))]
+pub struct Widgets<T: Widget>(HashMap<String, T>);
```

Comment:
type Widgets<K, W> = HashMap<K, W>;
type ButtonWidgets<K> = Widgets<K, UiButton>;

Reply:
I've changed this so that the widget id type is now an associated type on `Widget`. The problem with this approach is the following:
Imagine there is an external library which adds lots of UI components. They define ButtonWidgets as `Widgets<String, UiButton>`. Now you have already created a few of your own button widgets before, but you used them as `Widgets<u32, UiButton>`. Now there are 2 instances, and it's possible that libraries conflict between each other or with your code in this way.